zerver/tests/test_signup.py

```diff
@@ -3232,6 +3232,62 @@ def test_create_realm_with_marketing_emails_enabled(self) -> None:
         self.assertEqual(user.realm, realm)
         self.assertTrue(user.enable_marketing_emails)
 
+    @override_settings(OPEN_REALM_CREATION=True, CORPORATE_ENABLED=False)
+    def test_create_realm_without_prompting_for_marketing_emails(self) -> None:
+        password = "test"
+        string_id = "zuliptest"
+        email = "user1@test.com"
+        realm_name = "Test"
+
+        # Make sure the realm does not exist
+        with self.assertRaises(Realm.DoesNotExist):
+            get_realm(string_id)
+
+        # Create new realm with the email
+        result = self.client_post("/new/", {"email": email})
+        self.assertEqual(result.status_code, 302)
+        self.assertTrue(result["Location"].endswith(f"/accounts/new/send_confirm/{email}"))
+        result = self.client_get(result["Location"])
+        self.assert_in_response("Check your email so we can get started.", result)
+
+        # Visit the confirmation link.
+        confirmation_url = self.get_confirmation_url_from_outbox(email)
+        result = self.client_get(confirmation_url)
+        self.assertEqual(result.status_code, 200)
+
+        # Simulate the initia POST that is made by confirm-preregistration.js
+        # by triggering submit on confirm_preregistration.html.
+        payload = {
+            "full_name": "",
+            "key": find_key_by_email(email),
+            "from_confirmation": "1",
+        }
+        result = self.client_post("/accounts/register/", payload)
+        # Assert that the form did not prompt the user for enabling
+        # marketing emails.
+        self.assert_not_in_success_response(
+            ["Subscribe me to Zulip's low-traffic newsletter"], result
+        )
+
+        result = self.submit_reg_form_for_user(
+            email,
+            password,
+            realm_subdomain=string_id,
+            realm_name=realm_name,
+        )
+        self.assertEqual(result.status_code, 302)
+
+        result = self.client_get(result.url, subdomain=string_id)
+        self.assertEqual(result.status_code, 302)
+        self.assertEqual(result.url, "http://zuliptest.testserver")
+
+        # Make sure the realm is created
+        realm = get_realm(string_id)
+        self.assertEqual(realm.string_id, string_id)
+        user = get_user(email, realm)
+        self.assertEqual(user.realm, realm)
+        self.assertTrue(user.enable_marketing_emails)
```

Comment:
Can we verify the value the user got for `user_profile.enable_marketing_emails` here?

Reply:
Beyond just seeing what `user_profile.enable_marketing_emails` is set to, I don't think so. However, you just pointed out a bug in my test where I should be using `assertFalse`, but instead used `assertTrue` by mistake.

Turns out that our test helper `submit_reg_form_for_user` passes in a default of `enable_marketing_emails=True` for all requests. An indirect way of making sure that the correct value is passed to `user_profile.enable_marketing_emails` is to not pass `enable_marketing_emails` in payloads where we don't expect the frontend to prompt the user for enabling marketing emails and then assert if `user_profile.enable_marketing_emails` is `False`. I'm going to update the PR to do this! Thanks for pointing this out! :)

Reply:
Yeah, that's pretty subtle.